File: src/report.jl
# Pure JLD2 reading/parsing for `FieldWriter` snapshot files. Deliberately
# Makie-free (unlike plotting.jl/HIT3DMakieExt.jl) so it lives in the
# package proper and is exercised by the ordinary test suite; the Makie
# extension calls these functions and only builds `Figure`s from the
# result — it does no JLD2 parsing of its own.

using .Schema: is_stepkey, stepnum, gridkey, serieskey
using JLD2: jldopen

"""Sorted `"step_########"` group names in an open `file`."""
_stepkeys(file) = sort(filter(is_stepkey, keys(file)))

"""
    _read_grid(file, T) -> (grid, ν)

Reconstruct the `Grid` (CPU, precision `T`) and read ν from a `FieldWriter`
file's self-describing `grid` group. Falls back to a 2π cube with unknown ν
for files predating the metadata group, inferring `Nx` from the rfft layout
of the first snapshot (assumes even `Nx`).
"""
function _read_grid(file, T)
    if haskey(file, "grid")
        g = Grid(file[gridkey(:Nx)], file[gridkey(:Ny)], file[gridkey(:Nz)];
                 Lx = file[gridkey(:Lx)], Ly = file[gridkey(:Ly)],
                 Lz = file[gridkey(:Lz)], T = T)
        ν = haskey(file, gridkey(:nu)) ? file[gridkey(:nu)] : nothing
        return g, ν
    end
    keys_ = _stepkeys(file)
    isempty(keys_) && error("no snapshots found in file")
    û = file[first(keys_) * "/û"]
    Nx = 2 * (size(û, 1) - 1)   # rfft layout; assumes even Nx
    @warn "no grid metadata in file; assuming a 2π cube (N = $Nx) and unknown ν"
    return Grid(Nx, size(û, 2), size(û, 3); T = T), nothing
end

"""Precision of the stored state, inferred from the first snapshot's `û`."""
_state_T(file) = real(eltype(file[first(_stepkeys(file)) * "/û"]))

"""
    load_run(path) -> (; grid, ν, T, steps)

Open the `FieldWriter` file at `path` and return its self-describing grid,
viscosity (`nothing` if the file has no `grid/nu` entry), state precision,
and the sorted vector of snapshot step numbers. Cheap: does not load any
field arrays. Errors if the file has no snapshots.
"""
function load_run(path)
    jldopen(path, "r") do file
        keys_ = _stepkeys(file)
        isempty(keys_) && error("no snapshots found in $path")
        T = _state_T(file)
        g, ν = _read_grid(file, T)
        return (; grid = g, ν, T, steps = stepnum.(keys_))
    end
end

"""
    read_series(path_or_file, name) -> NamedTuple or nothing

Read the dense `series/<name>` group written by `save_series` — columnar:
`t` plus either `values` (scalar diagnostics) or one vector per
named-tuple field (e.g. `E`, `ε`, `P` for `energy_budget`) — as a
`NamedTuple`, or `nothing` if the file has no such group.
"""
function read_series(path::AbstractString, name)
    jldopen(path, "r") do file
        read_series(file, name)
    end
end

function read_series(file, name)
    grp = serieskey(name)
    haskey(file, grp) || return nothing
    fields = [k for k in keys(file[grp]) if k != "t"]
    pairs = Pair{Symbol, Any}[:t => file[grp * "/t"]]
    for fld in fields
        push!(pairs, Symbol(fld) => file[grp * "/" * fld])
    end
    return NamedTuple(pairs)
end

File: src/hit3d/report.py
"""Reading/parsing for `FieldWriter` snapshot files (HDF5).

Deliberately matplotlib-free (unlike `plotting`), so it is exercised by the
ordinary test suite; the plotting functions call these and only build
figures from the result — they do no HDF5 parsing of their own.
"""

from __future__ import annotations

import warnings
from types import SimpleNamespace

import h5py
import numpy as np

from .grid import make_grid
from .schema import grid_key, is_step_key, series_key, step_key, step_num

__all__ = ["load_run", "read_series", "step_keys", "read_grid", "read_snapshot"]


def step_keys(f):
    """Sorted ``step_########`` group names in an open h5py file."""
    return sorted(k for k in f.keys() if is_step_key(k))


def read_grid(f, dtype):
    """Reconstruct the `Grid` (precision `dtype`) and read ν from a
    `FieldWriter` file's self-describing `grid` group. Falls back to a 2π
    cube with unknown ν for files predating the metadata group, inferring
    the sizes from the rfft layout of the first snapshot (assumes even Nz).
    """
    if "grid" in f:
        g = make_grid(
            int(f[grid_key("Nx")][()]),
            int(f[grid_key("Ny")][()]),
            int(f[grid_key("Nz")][()]),
            Lx=float(f[grid_key("Lx")][()]),
            Ly=float(f[grid_key("Ly")][()]),
            Lz=float(f[grid_key("Lz")][()]),
            dtype=dtype,
        )
        nu = float(f[grid_key("nu")][()]) if grid_key("nu") in f else None
        return g, nu
    keys = step_keys(f)
    if not keys:
        raise ValueError("no snapshots found in file")
    shape = f[f"{keys[0]}/u_hat"].shape  # (3, Nx, Ny, Nz//2+1)
    nz = 2 * (shape[3] - 1)
    warnings.warn(
        f"no grid metadata in file; assuming a 2π cube "
        f"({shape[1]}x{shape[2]}x{nz}) and unknown nu",
        stacklevel=2,
    )
    return make_grid(shape[1], shape[2], nz, dtype=dtype), None


def _state_dtype(f):
    """Real precision of the stored state, from the first snapshot."""
    c = f[f"{step_keys(f)[0]}/u_hat"].dtype
    return np.float64 if c == np.complex128 else np.float32


def read_snapshot(f, step):
    """Spectral state of snapshot `step` from an open file, as NumPy."""
    return np.asarray(f[f"{step_key(step)}/u_hat"])


def load_run(path):
    """Open the `FieldWriter` file at `path` and return a namespace with
    its self-describing ``grid``, ``nu`` (None if the file has no grid/nu
    entry), state ``dtype``, and the sorted list of snapshot ``steps``.
    Cheap: does not load any field arrays. Errors if the file has no
    snapshots."""
    with h5py.File(path, "r") as f:
        keys = step_keys(f)
        if not keys:
            raise ValueError(f"no snapshots found in {path}")
        dtype = _state_dtype(f)
        grid, nu = read_grid(f, dtype)
        return SimpleNamespace(
            grid=grid, nu=nu, dtype=dtype, steps=[step_num(k) for k in keys]
        )


def read_series(path_or_file, name):
    """Read the dense ``series/<name>`` group written by `save_series` —
    columnar: `t` plus either `values` (scalar diagnostics) or one vector
    per budget field (e.g. E, eps, P) — as a dict of NumPy arrays, or None
    if the file has no such group."""
    if isinstance(path_or_file, h5py.File):
        return _read_series(path_or_file, name)
    with h5py.File(path_or_file, "r") as f:
        return _read_series(f, name)


def _read_series(f, name):
    grp = series_key(name)
    if grp not in f:
        return None
    return {k: np.asarray(f[grp][k]) for k in f[grp].keys()}
